File: happy.ts
// ============================================================
// ADVANCED TYPESCRIPT · HAPPY NUMBER · TYPE-SAFE LUXE
// ============================================================
// domain: number theory, functional programming, type predicates
// references: Pierce, B.C. (2002). Types and Programming Languages.
//             Hutton, G. (2016). Programming in Haskell (2nd ed.).
// ============================================================

type HappyResult = {
  readonly number: number;
  readonly isHappy: boolean;
  readonly trace: readonly number[];
  readonly cycleDetected: boolean;
};

type Statistics = {
  readonly limit: number;
  readonly total: number;
  readonly happyCount: number;
  readonly unhappyCount: number;
  readonly density: number;
  readonly happySample: readonly number[];
};

class HappyNumberAnalyzer {
  private readonly cache: Map<number, boolean> = new Map();

  /**
   * Compute sum of squares of digits without string conversion.
   * O(log n) time, O(1) space.
   */
  private static sumOfSquares(n: number): number {
    let total = 0;
    let num = n;
    while (num > 0) {
      const digit = num % 10;
      total += digit * digit;
      num = Math.floor(num / 10);
    }
    return total;
  }

  /**
   * Core algorithm with type-safe result.
   * Uses Set for O(1) lookups during cycle detection.
   */
  public isHappy(n: number): boolean {
    if (n <= 0 || !Number.isInteger(n)) {
      return false;
    }

    // cache hit
    if (this.cache.has(n)) {
      return this.cache.get(n)!;
    }

    const seen = new Set<number>();
    let current = n;

    while (current !== 1 && !seen.has(current)) {
      seen.add(current);
      current = HappyNumberAnalyzer.sumOfSquares(current);
    }

    const result = current === 1;
    this.cache.set(n, result);
    return result;
  }

  /**
   * Returns detailed analysis including the full trace.
   */
  public analyze(n: number): HappyResult {
    if (n <= 0 || !Number.isInteger(n)) {
      return {
        number: n,
        isHappy: false,
        trace: [],
        cycleDetected: false,
      };
    }

    const seen = new Set<number>();
    const trace: number[] = [];
    let current = n;
    let cycleDetected = false;

    while (current !== 1 && !seen.has(current)) {
      seen.add(current);
      trace.push(current);
      current = HappyNumberAnalyzer.sumOfSquares(current);
    }

    trace.push(current);
    if (current !== 1) {
      cycleDetected = true;
    }

    return {
      number: n,
      isHappy: current === 1,
      trace: trace as readonly number[],
      cycleDetected,
    };
  }

  /**
   * Batch process multiple numbers efficiently.
   */
  public batchCheck(numbers: readonly number[]): readonly HappyResult[] {
    return numbers.map((n) => this.analyze(n));
  }

  /**
   * Generate statistical summary for range [1, limit].
   */
  public statistics(limit: number): Statistics {
    if (limit < 1) {
      throw new Error('Limit must be >= 1');
    }

    const happyNumbers: number[] = [];
    for (let i = 1; i <= limit; i++) {
      if (this.isHappy(i)) {
        happyNumbers.push(i);
      }
    }

    return {
      limit,
      total: limit,
      happyCount: happyNumbers.length,
      unhappyCount: limit - happyNumbers.length,
      density: happyNumbers.length / limit,
      happySample: happyNumbers.slice(0, 20) as readonly number[],
    };
  }

  /**
   * Clear the internal cache to free memory.
   */
  public clearCache(): void {
    this.cache.clear();
  }
}

// ============================================================
// FUNCTIONAL EXPORT (compatible with test harness)
// ============================================================

export function happy(n: number): boolean {
  const analyzer = new HappyNumberAnalyzer();
  return analyzer.isHappy(n);
}

// ============================================================
// SELF-TEST & DEMONSTRATION
// ============================================================

if (require.main === module) {
  const analyzer = new HappyNumberAnalyzer();

  const testData: [number, boolean][] = [
    [1, true],
    [2, false],
    [7, true],
    [10, true],
    [13, true],
    [19, true],
    [23, true],
    [28, true],
    [31, true],
    [32, true],
    [33, false],
  ];

  console.log('🧪 HAPPY NUMBER TYPE-SAFE TEST SUITE');
  console.log('='.repeat(50));

  let allPassed = true;
  for (const [num, expected] of testData) {
    const result = analyzer.isHappy(num);
    const analysis = analyzer.analyze(num);
    const status = result === expected ? '✅' : '❌';
    const traceStr = analysis.trace.slice(0, 5).join(' → ');
    const suffix = analysis.trace.length > 5 ? ' …' : '';
    console.log(`${status} happy(${num}) → ${result} (expected ${expected})  trace: ${traceStr}${suffix}`);
    if (result !== expected) allPassed = false;
  }

  console.log('='.repeat(50));
  console.log(`OVERALL: ${allPassed ? '✅ ALL TESTS PASSED' : '❌ SOME TESTS FAILED'}`);

  const stats = analyzer.statistics(100);
  console.log('\n📊 STATISTICAL ANALYSIS (1–100)');
  console.log(`  Density: ${(stats.density * 100).toFixed(2)}%`);
  console.log(`  First 10 happy numbers: ${stats.happySample.slice(0, 10).join(', ')}`);

  // performance benchmark
  const start = performance.now();
  for (let i = 1; i <= 10000; i++) {
    analyzer.isHappy(i);
  }
  const elapsed = performance.now() - start;
  console.log(`\n⚡ Performance: checked 10,000 numbers in ${elapsed.toFixed(2)}ms`);
}

File: happy.py
# ============================================================
# ADVANCED PYTHON · HAPPY NUMBER · LUXE EDITION
# ============================================================
# domain: number theory, deterministic state-machine, cycle detection
# references: Flajolet, P., & Sedgewick, R. (2009). Analytic Combinatorics.
#             Knuth, D.E. (1997). The Art of Computer Programming, Vol.2.
# ============================================================

from typing import Union, List, Tuple, Optional
from functools import lru_cache
import math
import sys


class HappyNumberEngine:
    """
    A high-performance, memoized happy number verifier with trace generation.
    Implements Floyd's cycle-detection (tortoise-hare) for O(λ+μ) time,
    but we use set-based for clarity and traceability.
    """

    def __init__(self, cache_size: int = 10_000):
        """
        Initialize the engine with an LRU cache for repeated queries.
        :param cache_size: maximum number of results to cache.
        """
        self.cache_size = cache_size
        self._cache = {}

    @staticmethod
    def _sum_of_squares(n: int) -> int:
        """
        Compute sum of squares of digits in O(log n) time.
        Inlined for speed; no string conversion to avoid overhead.
        """
        total = 0
        while n:
            digit = n % 10
            total += digit * digit
            n //= 10
        return total

    def is_happy(self, n: int) -> bool:
        """
        Returns True if n is a happy number, False otherwise.
        Uses set-based detection with optional caching.
        """
        if n <= 0:
            return False

        # cache lookup
        if n in self._cache:
            return self._cache[n]

        seen = set()
        current = n
        while current != 1 and current not in seen:
            seen.add(current)
            current = self._sum_of_squares(current)

        result = (current == 1)
        # store in cache (only if not too large)
        if len(self._cache) < self.cache_size:
            self._cache[n] = result
        return result

    def get_trace(self, n: int) -> List[int]:
        """
        Returns the full trace of the happy-number process until termination
        or cycle detection. Useful for visualization and debugging.
        """
        if n <= 0:
            return []

        seen = set()
        trace = []
        current = n
        while current != 1 and current not in seen:
            seen.add(current)
            trace.append(current)
            current = self._sum_of_squares(current)
        trace.append(current)  # terminal state (1 or cycle entry)
        return trace

    def batch_check(self, numbers: List[int]) -> List[Tuple[int, bool]]:
        """
        Efficiently check multiple numbers, reusing cache.
        """
        return [(num, self.is_happy(num)) for num in numbers]

    def statistical_analysis(self, limit: int = 1000) -> dict:
        """
        Compute density and distribution of happy numbers up to limit.
        Returns a dictionary with statistics.
        """
        if limit < 1:
            return {"error": "limit must be >= 1"}

        happy_count = 0
        happy_list = []
        for i in range(1, limit + 1):
            if self.is_happy(i):
                happy_count += 1
                happy_list.append(i)

        return {
            "limit": limit,
            "total_numbers": limit,
            "happy_count": happy_count,
            "unhappy_count": limit - happy_count,
            "density": happy_count / limit,
            "happy_numbers_sample": happy_list[:20],  # first 20 for brevity
        }

    def clear_cache(self) -> None:
        """Clear the internal result cache."""
        self._cache.clear()


# ============================================================
# FUNCTIONAL INTERFACE (compatibility with tests)
# ============================================================

def happy(n: int) -> bool:
    """
    Standalone function that returns True if n is happy.
    Uses a global engine instance for caching.
    """
    _global_engine = HappyNumberEngine()
    return _global_engine.is_happy(n)


# ============================================================
# DEMONSTRATION & SELF-TEST
# ============================================================

if __name__ == "__main__":
    # test cases from problem statement
    test_cases = [
        (1, True),
        (2, False),
        (7, True),
        (10, True),
        (13, True),
        (19, True),
        (23, True),
        (28, True),
        (31, True),
        (32, True),
        (33, False),
    ]

    engine = HappyNumberEngine()
    print("🧪 HAPPY NUMBER VALIDATION SUITE")
    print("=" * 50)

    all_passed = True
    for num, expected in test_cases:
        result = engine.is_happy(num)
        trace = engine.get_trace(num)
        status = "✅" if result == expected else "❌"
        print(f"{status} happy({num}) → {result} (expected {expected})  trace: {trace[:5]}{'…' if len(trace) > 5 else ''}")
        if result != expected:
            all_passed = False

    print("=" * 50)
    print(f"OVERALL: {'✅ ALL TESTS PASSED' if all_passed else '❌ SOME TESTS FAILED'}")

    # statistical analysis
    stats = engine.statistical_analysis(100)
    print("\n📊 STATISTICAL ANALYSIS (1–100)")
    print(f"  Density of happy numbers: {stats['density']:.2%}")
    print(f"  First 10 happy numbers: {stats['happy_numbers_sample'][:10]}")

    # performance test
    import time
    start = time.perf_counter()
    _ = [engine.is_happy(i) for i in range(1, 10001)]
    elapsed = time.perf_counter() - start
    print(f"\n⚡ Performance: checked 10,000 numbers in {elapsed:.4f}s")
